src/main/java/com/openlattice/datastore/services/BackgroundIndexingService.java

```diff
@@ -0,0 +1,147 @@
+package com.openlattice.datastore.services;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+import com.google.common.collect.SetMultimap;
+import com.google.common.collect.Sets;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.IMap;
+import com.openlattice.conductor.rpc.ConductorElasticsearchApi;
+import com.openlattice.data.storage.PostgresEntityDataQueryService;
+import com.openlattice.edm.EntitySet;
+import com.openlattice.edm.type.EntityType;
+import com.openlattice.edm.type.PropertyType;
+import com.openlattice.hazelcast.HazelcastMap;
+import com.openlattice.neuron.audit.AuditEntitySetUtils;
+import com.openlattice.postgres.DataTables;
+import com.openlattice.postgres.PostgresColumn;
+import com.openlattice.postgres.ResultSetAdapters;
+import com.zaxxer.hikari.HikariDataSource;
+import org.apache.olingo.commons.api.edm.EdmPrimitiveTypeKind;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.scheduling.annotation.Scheduled;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+public class BackgroundIndexingService {
+    private static final Logger logger     = LoggerFactory.getLogger( BackgroundIndexingService.class );
+    private static final int    BLOCK_SIZE = 10000;
+
+    private final HikariDataSource               hds;
+    private final PostgresEntityDataQueryService dataQueryService;
+    private final ConductorElasticsearchApi      elasticsearchApi;
+
+    private final IMap<UUID, PropertyType> propertyTypes;
+    private final IMap<UUID, EntityType>   entityTypes;
+    private final IMap<UUID, EntitySet>    entitySets;
+
+    private final String getDirtyEntities;
+
+    public BackgroundIndexingService(
+            HikariDataSource hds,
+            HazelcastInstance hazelcastInstance,
+            ConductorElasticsearchApi elasticsearchApi,
+            PostgresEntityDataQueryService dataQueryService ) {
+        this.hds = hds;
+        this.dataQueryService = dataQueryService;
+        this.elasticsearchApi = elasticsearchApi;
+
+        this.propertyTypes = hazelcastInstance.getMap( HazelcastMap.PROPERTY_TYPES.name() );
+        this.entityTypes = hazelcastInstance.getMap( HazelcastMap.ENTITY_TYPES.name() );
+        this.entitySets = hazelcastInstance.getMap( HazelcastMap.ENTITY_SETS.name() );
+
+        indexUpdatedEntitySets();
+
+    }
+
+    private String getDirtyEntitiesQuery( UUID entitySetId ) {
+        return "SELECT * FROM " + DataTables.entityTableName( entitySetId ) + " WHERE "
+                + PostgresColumn.LAST_INDEX_FIELD + " < " + PostgresColumn.LAST_WRITE_FIELD + " LIMIT " + String
+                .valueOf( BLOCK_SIZE )
+                + " OFFSET ?";
```

Comment:
You want to use limit instead of offset. Off set will skip block size rows instead of limiting response to block_size rows.